src/Build/Evaluation/ItemSpec.cs

```diff
@@ -85,6 +87,40 @@ public override bool IsMatch(string itemToMatch)
                 return ReferencedItems.Any(v => v.ItemAsValueFragment.IsMatch(itemToMatch));
             }
 
+            public override bool IsMatchOnMetadata(IItem item, IEnumerable<string> metadata, MatchOnMetadataOptions options)
+            {
+                switch (options)
+                {
+                    case MatchOnMetadataOptions.CaseSensitive:
+                        return IsCaseSensitiveMetadataMatch(item, metadata, ReferencedItems);
+                    case MatchOnMetadataOptions.CaseInsensitive:
+                        return IsCaseInsensitiveMetadataMatch(item, metadata, ReferencedItems);
+                    case MatchOnMetadataOptions.PathLike:
+                        return IsPathLikeMetadataMatch(item, metadata, ReferencedItems, ProjectDirectory);
+                    default:
+                        ErrorUtilities.ThrowInternalErrorUnreachable();
+                        return false;
+                }
+            }
+
+            private static bool IsCaseSensitiveMetadataMatch(IItem item, IEnumerable<string> metadata, List<ReferencedItem> referencedItems)
+            {
+                return referencedItems.Any(referencedItem =>
+                            metadata.All(m => item.GetMetadataValue(m).Equals(referencedItem.Item.GetMetadataValue(m)) && !item.GetMetadataValue(m).Equals(string.Empty)));
+            }
+
+            private static bool IsCaseInsensitiveMetadataMatch(IItem item, IEnumerable<string> metadata, List<ReferencedItem> referencedItems)
+            {
+                return referencedItems.Any(referencedItem =>
+                    metadata.All(m => String.Equals(item.GetMetadataValue(m), referencedItem.Item.GetMetadataValue(m), StringComparison.OrdinalIgnoreCase) && !item.GetMetadataValue(m).Equals(string.Empty)));
+            }
+
+            private static bool IsPathLikeMetadataMatch(IItem item, IEnumerable<string> metadata, List<ReferencedItem> referencedItems, string projectDirectory)
+            {
+                return referencedItems.Any(referencedItem =>
+                    metadata.All(m => !item.GetMetadataValue(m).Equals(string.Empty) && FileUtilities.ComparePathsNoThrow(item.GetMetadataValue(m), referencedItem.Item.GetMetadataValue(m), projectDirectory)));
+            }
+
```

Comment:
Here's how I would unify them.
```suggestion
bool MetadataComparer(string itemMetadata, string referencedItemMetadata) => options.Equals(MatchOnMetadataOptions.PathLike) ? FileUtilities.ComparePathsNoThrow(itemMetadata, referencedItemMetadata, ProjectDirectory) :
                String.Equals(itemMetadata, referencedItemMetadata, options.Equals(MatchOnMetadataOptions.CaseInsensitive) ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal);

                return ReferencedItems.Any(referencedItem =>
                        metadata.All(m => !item.GetMetadataValue(m).Equals(string.Empty) && MetadataComparer(item.GetMetadataValue(m), referencedItem.Item.GetMetadataValue(m))));
            }
```

Reply:
But you can also take the MetadataComparer (or other name) and put it in a helper method; it would just need one more parameter.

Reply:
As another option:
```suggestion
                bool MetadataComparer(string itemMetadata, string referencedItemMetadata)
                {
                    switch (options)
                    {
                        case MatchOnMetadataOptions.CaseInsensitive:
                            return String.Equals(itemMetadata, referencedItemMetadata, StringComparison.OrdinalIgnoreCase);
                        case MatchOnMetadataOptions.CaseSensitive:
                            return String.Equals(itemMetadata, referencedItemMetadata, StringComparison.Ordinal);
                        case MatchOnMetadataOptions.PathLike:
                            return FileUtilities.ComparePathsNoThrow(itemMetadata, referencedItemMetadata, ProjectDirectory);
                        default:
                            ErrorUtilities.ThrowInternalErrorUnreachable();
                            return false;
                    }
                }

                return ReferencedItems.Any(referencedItem =>
                        metadata.All(m => !item.GetMetadataValue(m).Equals(string.Empty) && MetadataComparer(item.GetMetadataValue(m), referencedItem.Item.GetMetadataValue(m))));
            }
```

I like this less because I like ? : more than switch, but it still fixes most of the issue to my mind. Even better: unify CaseSensitive and CaseInsensitive with a ?: operator and make them the default.

Reply:
I'm pretty much neutral on all these options. @rainersigwald do you have strong feelings?

Reply:
I'd say to leave it the way it is now. I don't find the reduced-duplication options clearer, and they have runtime perf characteristics I don't fully understand (since they use a local function with a captured variable from the outer scope.

Reply:
The middle option (putting the helper function as a separate function taking two strings and a MatchOnMetadataOptions as inputs) wouldn't have that as a potential problem, and it would reduce the code duplication. It also makes it clearer that this is three ways of comparing the metadata of an item with that of the referenced item.